README
NAME
    Pod::WikiDoc - Generate Pod from inline wiki style text

VERSION
    This documentation refers to version 0.19.

SYNOPSIS
    In a source file, Pod format-block style:

         =begin wikidoc
     
     = POD FORMAT-BLOCK STYLE
     
     Write documentation with *bold*, ~italic~ or {code}
         markup.  Create a link to [Pod::WikiDoc].
         Substitute for user-defined %%KEYWORD%%.
     
         Indent for verbatim paragraphs
     
     * bullet
         * point
         * list
     
     0 sequentially
         0 numbered
         0 list
     
     = end wikidoc

    In a source file, wikidoc comment-block style:

         ### = WIKIDOC COMMENT-BLOCK STYLE
         ###
         ### Optionally, [Pod::WikiDoc] can extract from
         ### specially-marked comment blocks

    Generate Pod from wikidoc, programmatically:

         use Pod::WikiDoc;
         my $parser = Pod::WikiDoc->new( { 
             comment_blocks => 1,
             keywords => { KEYWORD => "foo" },
         } );
         $parser->filter( 
             { input => "my_module.pm", output => "my_module.pod" }
         );

    Generate Pod from wikidoc, via command line:

         $ wikidoc -c my_module.pm my_module.pod

DESCRIPTION
    Pod works well, but writing it can be time-consuming and tedious. For
    example, commonly used layouts like lists require numerous lines of text
    to make just a couple of simple points. An alternative approach is to
    write documentation in a wiki-text shorthand (referred to here as
    *wikidoc*) and use Pod::WikiDoc to extract it and convert it into its
    corresponding Pod as a separate ".pod" file.

    Documentation written in wikidoc may be embedded in Pod format blocks,
    or, optionally, in specially marked comment blocks. Wikidoc uses simple
    text-based markup like wiki websites to indicate formatting and links.
    (See "WIKIDOC MARKUP", below.)

    Pod::WikiDoc processes text files (or text strings) by extracting both
    existing Pod and wikidoc, converting the wikidoc to Pod, and then
    writing the combined document back to a file or standard output.

    Summary of major features of Pod::WikiDoc:

    *   Extracts and converts wikidoc from Pod format blocks or special
        wikidoc comment blocks

    *   Extracts and preserves existing Pod

    *   Provides bold, italic, code, and link markup

    *   Substitutes user-defined keywords

    *   Automatically converts special symbols in wikidoc to their Pod
        escape equivalents, e.g. E<lt>, E<gt>

    *   Preserves other Pod escape sequences, e.g. E<euro>

    In addition, Pod::WikiDoc provides a command-line utility, wikidoc, to
    simplify wikidoc translation.

    See the Pod::WikiDoc::Cookbook for more detailed usage examples,
    including how to automate ".pod" generation when using Module::Build.

INTERFACE
  "new"
         $parser = Pod::WikiDoc->new( \%args );

    Constructor for a new Pod::WikiDoc object. It takes a single, optional
    argument: a hash reference with the following optional keys:

    *   "comment_blocks": if true, Pod::WikiDoc will scan for wikidoc in
        comment blocks. Default is false.

    *   "comment_prefix_length": the number of leading sharp (#) symbols to
        denote a comment block. Default is 3.

    *   "keywords": a hash reference with keywords and values for keyword
        substitution == "convert"

         my $pod_text = $parser->convert( $input_text );

    Given a string with valid Pod and/or wikidoc markup, filter/translate it
    to Pod. This is really just a wrapper around "filter" for working with
    strings rather than files, and provides similar behavior, including
    adding a 'Generated by' header. == "filter"

         $parser->filter( \%args );

    Filters from an input file for Pod and wikidoc, translating it to Pod
    and writing it to an output file. The output file will be prefixed with
    a 'Generated by' comment with the version of Pod::WikiDoc and timestamp,
    as required by perlpodspec.

    "filter" takes a single, optional argument: a hash reference with the
    following optional keys:

    *   "input": a filename or filehandle to read from. Defaults to STDIN.

    *   "output": a filename or filehandle to write to. If given a filename
        and the file already exists, it will be clobbered. Defaults to
        STDOUT. == "format"

         my $pod_text = $parser->format( $wiki_text );

    Given a string with valid Pod and/or wikidoc markup, filter/translate it
    to Pod. Unlike "convert", no 'Generated by' comment is added. This
    function is used internally by Pod::WikiDoc, but is being made available
    as a public method for users who want more granular control of the
    translation process or who want to convert wikidoc to Pod for other
    creative purposes using the Pod::WikiDoc engine.

WIKIDOC MARKUP
    Pod::WikiDoc uses a wiki-style text markup, called wikidoc. It is
    heavily influenced by Kwiki. Like other wiki markup, it has both block
    and inline elements, which map directly to their Pod equivalents.

    Block elements include:

    *   Headers

    *   Verbatim text

    *   Bullet lists

    *   Numbered lists

    *   Ordinary paragraphs

    Block elements should be separated by a blank line (though Pod::WikiDoc
    will do the right thing in many cases if you don't).

    Inline elements include:

    *   Bold

    *   Italic

    *   Code

    *   Link

    *   Escape code

    *   Keywords

    All text except that found in verbatim text, code markup or keywords is
    transformed to convert special Pod characters to Pod escape code markup:
    E<lt>, E<gt>, E<sol>, E<verbar>. Inline markup can be escaped with a
    backslash (\). Including a literal backslash requires a double-backslash
    (\\).

  Headers
    Headers are indicated with one or more equals signs followed by
    whitespace in the first column. The number of equals signs indicates the
    level of the header (the maximum is four). Headers can not span multiple
    lines.

         = header level 1
     
     == header level 2

  Verbatim text
    Verbatim text is indicated with leading whitespace in each line of text,
    just as with Pod.

         #<--- first column 
     
         sub verbatim {}

  Bullet lists
    Bullet lists are indicated with an asterisk in the first column followed
    by whitespace. Bullet lists can span multiple lines. Lines after the
    first should not have an asterisk or be indented.

         * First item in the list
         * Second item in the list
         on multiple lines
         * Third item in the list

  Numbered lists
    Numbered lists work just like numbered lists, but with a leading 0
    followed by whitespace.

         0 First item in the list
         0 Second item in the list
         on multiple lines
         0 Third item in the list

  Ordinary paragraphs
    Ordinary paragraphs consist of one or more lines of text that do not
    match the criteria of other blocks. Paragraphs are terminated with a
    empty line.

         This is an ordinary paragraph that
         spans multiple lines.

  Bold markup
    Bold text is indicated by bracketing with asterisks. Bold markup must
    begin at a whitespace boundary, the start of a line, or the inside of
    other markup.

         This shows *bold* text.

  Italic markup
    Italic text is indicated by bracketing with tildes. Italic markup must
    begin at a whitespace boundary, the start of a line, or the inside of
    other markup.

         This shows ~italic~ text.

  Code markup
    Code (monospaced) text is indicated by bracketing with matched braces.
    Code markup must begin at a whitespace boundary, the start of a line, or
    the inside of other markup. Brackets should nest properly with code.

         This shows {code} text.  It can surround text
         with brackets like this: { $data{ $id } }

  Link markup
    Link text is indicated by bracketing with square brackets. As with Pod,
    link text may include a vertical bar to separate display text from the
    link itself. Link markup must begin at a whitespace boundary, the start
    of a line, or the inside of other markup.

         This is an ordinary [Pod::WikiDoc] link.
         This is a [way to ~markup~ links|Pod::WikiDoc] with display text
         Hypertext links look like this: [http://www.google.com/]

  Escape code markup
    Pod-style escape text is passed through as normal to support
    international or other unusual characters.

         This is the euro symbol: E<euro>

  Keyword markup
    Text surrounded by double-percent signs is treated as a keyword for
    expansion. The entire expression will be replaced with the value of the
    keyword from the hash provided when the parser is created with "new()".
    If the keyword is unknown or the value is undefined, the keyword will be
    passed through unchanged.

         This is version %%VERSION%%

DIAGNOSTICS
    *   "Error: Argument to convert() must be a scalar"

    *   "Error: Argument to filter() must be a hash reference"

    *   "Error: Argument to format() must be a scalar"

    *   "Error: Argument to new() must be a hash reference"

    *   "Error: Class method new() can't be called on an object"

    *   "Error: Couldn't open input file 'FILENAME'"

    *   "Error: Couldn't open output file 'FILENAME'"

    *   "Error: 'input' parameter for filter() must be a filename or
        filehandle"

    *   "Error: 'output' parameter for filter() must be a filename or
        filehandle"

INCOMPATIBILITIES
    *   Default prefix length for wikidoc comment-blocks conflicts with
        Smart::Comments. Change the "comment_prefix_length" argument to
        "new" in Pod::WikiDoc or the level of 'smartness' in Smart::Comments
        to avoid the conflict.

    *   Module::Build before 0.28 does not look in external ".pod" files to
        generate a "README" with the "create_readme" option or to find a
        module abstract. Set the abstract manually in the "Build.PL" file
        with the "dist_abstract" option.

BUGS
    Please report bugs or feature requests using the CPAN Request Tracker.
    Bugs can be submitted using the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Pod-WikiDoc>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2005, 2006, 2007 by David A. Golden

    Licensed under the Apache License, Version 2.0 (the "License"); you may
    not use this file except in compliance with the License. You may obtain
    a copy of the License at <http://www.apache.org/licenses/LICENSE-2.0>

    Files produced as output though the use of this software, including
    generated copies of boilerplate templates provided with this software,
    shall not be considered Derivative Works, but shall be considered the
    original work of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for Perl module Pod::WikiDoc

{{ $NEXT }}

0.21      2017-05-08 15:01:24-04:00 America/New_York

    - fixed test for perls without '.' in @INC

0.20      2012-02-29 12:29:20 EST5EDT

    - added dependencies that were ommitted from 0.19

0.19      2012-02-23 12:19:36 EST5EDT

    - attempted to fix problems caused by incompatible changes
      in Parse::RecDescent breaking pre-compiled parsers

    - converted distribution to Dist::Zilla management

0.18 Sun Sep 30 13:10:12 EDT 2007

    - rewrote Pod/wikidoc filtering; added support for nested formats
      (RT#29604 by ASKSH)
    - changed to the Apache License, version 2.0; (it's clearer, relicensable,
      and is explicit about contributions)
    - cleaned up some Perl::Critic level 4/5 warnings

0.17 Tue Mar 20 08:37:24 EDT 2007
    - forced numeric version comparison checks in 'Generated by' tests
      (e.g. 0.17 should be considered equal to 0.170)

0.16 Mon Dec  4 10:18:48 EST 2006
    - added support for parsing inside parentheses (~foo~)
    - fixed dependency lists in Makefile.PL and removed from Pod
    
0.15 Sun Jun 11 01:25:20 EDT 2006
    - fixed up version numbering across generated modules
    
0.14 Sun Jun 11 00:41:31 EDT 2006
    - added keyword expansion (e.g. %%VERSION%% )
    - added keyword definition options to wikidoc program
    - added no_index recipe to Cookbook for extra documentation files
    - added pod/pod-coverage tests

0.13 Mon Feb  6 22:52:16 EST 2006
    - Created custom Makefile.PL
    - Shifted parser compilation from build cycle back to distribution
      creation (less burden on end-user)

0.12 Sat Sep 24 09:17:56 EDT 2005
    - Removed Makefile.PL and using Build.PL only as the  Makefile can't
      support the dynamic build and doc process and users installing with it
      would see failures
    - Fixed dependency list in documentation
    - Various documentation tweaks and clarifications

0.11 Fri Sep 23 17:16:13 EDT 2005
    - Added Text::Balanced to Build.PL requires
    - Fixed typo in package name in Cookbook.pm
    - Tweaked doc phrasing for Pod::WikiDoc::Parser
    - Appended grammar to Pod::WikiDoc::Parser Pod file
    - Stubbed README (wasn't being generated properly)

0.10 Fri Sep 23 09:46:42 EDT 2005
    - basic wikidoc format section extraction
    - basic block formatting (header, paragraph, verbatim, lists)
    - basic inline formattting (bold, italic, link, code)
    - link formatting with or without link text
    - special symbols "<>|/" automatically escaped in generated Pod
    - format nesting
    - wikidoc comment-blocks ("### " indicates wikidoc)


t/CLI.pm
package t::CLI;
use strict;
use warnings;

use Test::Builder;
use Probe::Perl 0.01;
use Cwd qw( abs_path );
use IPC::Run3 0.033;
use File::Basename qw( basename );
use File::Spec 3;
#use POSIX qw( WEXITSTATUS );

my $Test = Test::Builder->new;
my $pp = Probe::Perl->new;
my $perl = abs_path($pp->find_perl_interpreter);
my $cwd = abs_path( "." );
my $coverdb = File::Spec->catdir($cwd,"cover_db");
my $cover = index( ($ENV{HARNESS_PERL_SWITCHES} || ''), '-MDevel::Cover' ) < 0
          ? ''
          : "-MDevel::Cover=-db,$coverdb"
          ;
#--------------------------------------------------------------------------#
# Main API
#--------------------------------------------------------------------------#

sub new {
    my ($class, $program, @default_args ) = @_;
    my $self = bless {}, $class;
    $self->program( $program );
    $self->default_args( [@default_args] );
    return $self;
}

# returns success or failure; exit_code is opposite (0 is good); actual error
# message is in @$
sub run {
    my ($self, @args) = @_;
    my ($stdout, $stderr);
    my $stdin = $self->stdin;

    my @cmd = (
        $perl, 
        ( -d 'blib' ? "-Mblib=$cwd" : "-Ilib=$cwd/lib" ), # must hard code this in case curdir changed
        ( $cover ? $cover : () ),
        $self->{program},
        @{ $self->default_args() },
        @args,
    );
                
    eval {
        run3 \@cmd, \$stdin, \$stdout, \$stderr;
        $self->exit_code( $? >> 8 || 0 );
    };
    my $result = $@ eq q{} ? 1 : 0;
    $self->stdout($stdout);
    $self->stderr($stderr);
    $@ .= "(running '@cmd')";
    return $result;
}

#--------------------------------------------------------------------------#
# Accessors
#--------------------------------------------------------------------------#

sub program {
    my ($self, $filename) = @_;
    if (defined $filename) {
        die "Can't find $filename" if ! -e $filename;
        $self->{program} = abs_path($filename);
    }
    return basename( $self->{program} );
}
        
BEGIN {
    my $evaltext = << 'CODE';
        sub {
            $_[0]->{PROP} = $_[1] if @_ > 1;
            return $_[0]->{PROP};
        }
CODE
    for ( qw( exit_code stdin stdout stderr default_args)) {
        no strict 'refs';
        (my $sub = $evaltext) =~ s/PROP/$_/g;
        *{__PACKAGE__ . "::$_"} = eval $sub;
    }
}

#--------------------------------------------------------------------------#
# Testing functions
#--------------------------------------------------------------------------#

sub runs_ok {
    my ($self, @args) = @_;
    my $label = "Ran " . $self->program() . " with "
              . (@args && $args[0] ne q{} ? "args '@args'" : "no args" )
              . " without error";
    my $runs = $self->run(@args);
    die $@ if ! $runs;
    my $ok = $Test->ok( ! $self->exit_code, $label );
    if ( ! $ok ) {
        $Test->diag( "Exit code: " . $self->exit_code . "\n");
        $Test->diag( "STDERR: " . $self->stderr . "\n") if $self->stderr;
        $Test->diag( "STDOUT: " . $self->stdout . "\n") if $self->stdout;
    }
    return $ok;
}

sub dies_ok {
    my ($self, @args) = @_;
    my $label = $self->program . " with "
              . (@args && $args[0] ne q{} ? "args '@args'" : "no args" )
              . " ended with an error";
    my $runs = $self->run(@args);
    die $@ if ! $runs;
    my $ok = $Test->ok( $self->exit_code, $label );
    if ( ! $ok ) {
        $Test->diag( "Exit code: " . $self->exit_code . "\n");
    }
    return $ok;
}


sub exits_with {
    my ($self, $expect, @args) = @_;
    my $label = $self->program . " with "
              . (@args && $args[0] ne q{} ? "args '@args'" : "no args" )
              . " ended with exit code $expect";
    my $runs = $self->run(@args);
    die $@ if ! $runs;
    return $Test->is_num( $self->exit_code, $expect, $label );
}

sub stdout_is { my ($self, $expect, $label) = @_; return $Test->is_eq(
        $self->stdout, $expect, $label ? "... $label" : "... " . $self->program
        . " had correct output to STDOUT" ); }

sub stdout_like {
    my ($self, $expect, $label) = @_;
    return $Test->like( $self->stdout, $expect, 
        $label 
            ? "... $label"
            : "... " . $self->program . " had correct output to STDOUT" );
}

sub stderr_is {
    my ($self, $expect, $label) = @_;
    return $Test->is_eq( $self->stderr, $expect,
        $label 
            ? "... $label"
            : "... " . $self->program . " had correct output to STDERR" );
}

sub stderr_like {
    my ($self, $expect, $label) = @_;
    return $Test->like( $self->stderr, $expect,
        $label 
            ? "... $label"
            : "... " . $self->program . " had correct output to STDERR" );
}


1; #true


t/Casefiles.pm
package t::Casefiles;
use strict;
use warnings;
use Carp;
use File::Find 1;
use File::Basename qw(basename);
use Test::More;

# Get optional test support
eval "use Test::Differences";
my $HAVE_DIFF = $@ eq '' ? 1 : 0;

#--------------------------------------------------------------------------#
# new
#--------------------------------------------------------------------------#

sub new {
    my ($class, $data_dir) = @_;

    # Verify directory
    croak "Error: Couldn't read data directory"
        if ( ! ( -e $data_dir && -r $data_dir ) );

    # create blesed object
    return bless { DATA_DIR => $data_dir }, $class;
}

#--------------------------------------------------------------------------#
# data_dir - accessor
#--------------------------------------------------------------------------#

sub data_dir { return $_[0]->{DATA_DIR} }

#--------------------------------------------------------------------------#
# files() - Return list of files
#--------------------------------------------------------------------------#

sub files {
    my ($self) = shift;
    my @case_files;

    my $options = { 
        wanted => sub {
            return if -d $File::Find::name;
            push @case_files, $File::Find::name;
        },
        no_chdir => 1,
    };

    File::Find::finddepth( $options, $self->data_dir );

    return @case_files;
}

#--------------------------------------------------------------------------#
# run_tests() - Run test for each case file
#--------------------------------------------------------------------------#

sub run_tests {
    my ($self, $filter) = @_;
    
    my @case_files = $self->files();

    if ( not Test::More->builder->has_plan ) {
        plan tests => scalar @case_files;
    }

    for my $case ( sort @case_files ) {
        open my $fh, "<", $case 
            or die "Couldn't open file $case";
        my $raw_data = do { local $/; <$fh> };
        my ($input, $expected) = split /^#-+#\n/ms, $raw_data;
        my $got = $filter->( $input );
        if ( $HAVE_DIFF ) {
            eq_or_diff( $got, $expected, basename( $case) );
        }
        else {
            is( $got, $expected, basename( $case ) );
        }
    }

}

1;


Build.PL
use Module::Build;

my $class = Module::Build->subclass(
    class => "Module::Build::WikiDoc",
    code => <<'SUBCLASS', );

    sub ACTION_wikidoc {
        my $self = shift;
        $self->depends_on('parser');
        eval "use Pod::WikiDoc";
        if ( $@ eq '' ) {
            my $parser = Pod::WikiDoc->new( { 
                comment_blocks => 1,
                keywords => { VERSION => $self->dist_version },
            });
            for my $src ( keys %{ $self->find_pm_files() } ) {
                (my $tgt = $src) =~ s{\.pm$}{.pod};
                $parser->filter( {
                    input   => $src,
                    output  => $tgt,
                });
                print "Creating $tgt\n";
                $tgt =~ s{\\}{/}g;
                $self->_add_to_manifest( 'MANIFEST', $tgt );
            }
        }
        else {
            warn "Pod::WikiDoc not available. Skipping wikidoc.\n";
        }
    }

    sub ACTION_testpod {
        my $self = shift;
        $self->depends_on('wikidoc');
        $self->SUPER::ACTION_testpod;
    }

    sub ACTION_parser {
        my $self = shift;
        my $parserfile = 'lib/Pod/WikiDoc/Parser.pm';
        $self->run_perl_script('Pod_WikiDoc_Parser.pm.PL', [],
                               [$parserfile] );
        $self->_add_to_manifest( 'MANIFEST', $parserfile );
    }

    sub ACTION_distdir {
        my $self = shift;
        $self->depends_on('parser');
        $self->depends_on('wikidoc');
        $self->SUPER::ACTION_distdir;
    }

#    sub ACTION_distmeta {
#        my $self = shift;
#        $self->meta_merge( 
#            provides => {
#                'Pod::WikiDoc::Parser' => {
#                    file => 'lib/Pod/WikiDoc/Parser.pm',
#                    version => $self->dist_version,
#                },
#            },
#        );
#        $self->SUPER::ACTION_distmeta;
#    }

SUBCLASS
            
$class->new( 
    module_name         => 'Pod::WikiDoc',
    dist_author         => 'David A Golden <dagolden@cpan.org>',
    dist_abstract       => 'Generate Pod from inline wiki style text',
    license             => 'apache',
    create_readme       => 1,
#    PL_files            => {
#        'Pod_WikiDoc_Parser.pm.PL' => 'lib/Pod/WikiDoc/Parser.pm',
#    },
#    add_to_cleanup      => [
#        'lib/Pod/WikiDoc/Parser.pod',
#    ],
    script_files => [
      'scripts/wikidoc',
    ],
    requires        => {
        perl                => 5.006,
        Getopt::Lucid       => 0.14,    
        IO::String          => 1.06,
        Parse::RecDescent   => 1.94,
        Pod::Usage          => 1,
        Scalar::Util        => 1.02,
        Text::Balanced      => 1.95,
        Cwd                 => 0,       #build_requires
        Test::More          => 0.45,    #build_requires
        IPC::Run3           => 0.033,   #build_requires
        Probe::Perl         => 0.01,    #build_requires
        File::Find          => 1,       #build_requires
        File::pushd         => 0.30,    #build_requires
        File::Spec          => 3,       #build_requires 
        File::Temp          => 0.14,    #build_requires
        File::Basename      => 0,       #build_requires
    },
    meta_add            => {
        no_index => {
            file => [ qw{
                lib/Pod/WikiDoc/Cookbook.pm
            } ],
            directory => [ 'examples' ],
        }
    },
)->create_build_script;



lib/Pod/WikiDoc/Cookbook.pm
package Pod::WikiDoc::Cookbook;
# Not really a .pm file, but holds wikidoc which will be
# turned into .pod by the Build.PL
use strict; # make CPANTS happy
use vars '$VERSION';
$VERSION = '0.19';
1;
__END__

=begin wikidoc

= NAME

Pod::WikiDoc::Cookbook - Examples of Pod::WikiDoc usage

= VERSION

This documentation refers to version %%VERSION%%.

= DESCRIPTION

This file contains some examples of ways to use [Pod::WikiDoc]
or to integrate [Pod::WikiDoc] with other tools.

= BASIC RECIPES

== Editor settings for working with Pod::WikiDoc

~(Seeking equivalent settings for other editors or alternatives for vim.)~

=== Vim

In vim, use the {comments} and {formatoptions} settings in {.vimrc} to
have vim automatically insert the wikidoc comment leader when pressing
return from a wikidoc comment line.  For example, the following lines in a 
{.vimrc} file will activate this option whenever a perl-ish file is loaded.
 
    autocmd BufNewFile,BufRead *.p? set comments=b:###
    autocmd BufNewFile,BufRead *.p? set formatoptions+=r

== Subclassing Module::Build to automatically generate Pod from wikidoc

With a little extra work in the Build.PL file, Pod::WikiDoc
can work easily with [Module::Build] to extract wikidoc into .pod
files automatically during the distribution process

The Build.PL file below subclasses [Module::Build] with three functions:

* ACTION_wikidoc -- adds a new {Build wikidoc} action that extracts Pod
and wikidoc from all .pm files in the {lib} directory and adds them to
the MANIFEST
* ACTION_testpod -- adds a dependency on the {wikidoc} action to regenerate
.pod files before testing them
* ACTION_distdir -- adds a dependency on the {wikidoc} action to regenerate
.pod files before bundling up a distribution

As an extra feature, ACTION_wikidoc also sets a VERSION keyword that can
be used to insert the current version number into the generated Pod.

    = VERSION
    This documentation refers to version %%VERSION%%.

By making wikidoc extraction part of the {distdir} action, users installing the
distribution will receive it with .pod files already created, and will not need
to have [Pod::WikiDoc] installed themselves.  

    # Build.PL
    use Module::Build;

    my $class = Module::Build->subclass(
        class => "Module::Build::WikiDoc",
        code => <<'SUBCLASS' );

        sub ACTION_wikidoc {
            my $self = shift;
            eval "use Pod::WikiDoc";
            if ( $@ eq '' ) {
                my $parser = Pod::WikiDoc->new( {
                    comment_blocks => 1}
                    keywords => { VERSION => $self->dist_version },
                });
                for my $src ( keys %{ $self->find_pm_files() } ) {
                    (my $tgt = $src) =~ s{\.pm$}{.pod};
                    $parser->filter( {
                        input   => $src,
                        output  => $tgt,
                    });
                    print "Creating $tgt\n";
                    $tgt =~ s{\\}{/}g; # for win32
                    $self->_add_to_manifest( 'MANIFEST', $tgt );
                }
            }
            else {
                warn "Pod::WikiDoc not available. Skipping wikidoc.\n";
            }
        }

        sub ACTION_testpod {
            my $self = shift;
            $self->depends_on('wikidoc');
            $self->SUPER::ACTION_testpod;
        }

        sub ACTION_distdir {
            my $self = shift;
            $self->depends_on('wikidoc');
            $self->SUPER::ACTION_distdir;
        }

    SUBCLASS

    $class->new( 
        # regular Module::Build options
    )->create_build_script;

== Including extra pure-wikidoc files in a distribution

To add extra documentation files to a distribution, create them as .pm
files and let Pod::WikiDoc convert them as normal.  To prevent the .pm
files from being indexed (e.g. by search.cpan.org), list them as "no_index"
in the META.yml file of the distribution.

Example of a simple .pm documentation file:

    package Some::Module::About;
    use strict; # make CPANTS happy
    1;
    __END__

    =begin wikidoc

    Your wikidoc goes here.
    
    =end wikidoc

Adding "no_index" to META.yml via Build.PL (requires Module::Build
0.28):

    my $builder = $class->new(
        # regular Module::Build options
        meta_add            => {
            no_index => {
                file => [ qw{
                    lib/Some/Module/About.pm
                } ]
            }
        },
    );

= AUTHOR

David A Golden (DAGOLDEN)

dagolden@cpan.org

[http://dagolden.com/]

= COPYRIGHT

Copyright (c) 2005,2006 by David A Golden

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=end wikidoc

=cut


Makefile.PL
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Generate Pod from inline wiki style text",
  "AUTHOR" => "David A Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Pod-WikiDoc",
  "EXE_FILES" => [
    "bin/wikidoc"
  ],
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Pod::WikiDoc",
  "PREREQ_PM" => {
    "Carp" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Basename" => 0,
    "File::Path" => 0,
    "File::pushd" => 0,
    "Getopt::Lucid" => "0.14",
    "IO::File" => 0,
    "IO::String" => "1.06",
    "Parse::RecDescent" => "1.967006",
    "Pod::Usage" => 1,
    "Scalar::Util" => "1.02",
    "Text::Balanced" => 0,
    "strict" => 0,
    "vars" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Carp" => 0,
    "Cwd" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Basename" => 0,
    "File::Find" => 1,
    "File::Spec" => 3,
    "File::Temp" => 0,
    "IO::String" => "1.06",
    "IPC::Run3" => "0.033",
    "Probe::Perl" => "0.01",
    "Test::Builder" => 0,
    "Test::More" => 0,
    "lib" => 0
  },
  "VERSION" => "0.22",
  "test" => {
    "TESTS" => "t/*.t t/contributed/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Cwd" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Basename" => 0,
  "File::Find" => 1,
  "File::Path" => 0,
  "File::Spec" => 3,
  "File::Temp" => 0,
  "File::pushd" => 0,
  "Getopt::Lucid" => "0.14",
  "IO::File" => 0,
  "IO::String" => "1.06",
  "IPC::Run3" => "0.033",
  "Parse::RecDescent" => "1.967006",
  "Pod::Usage" => 1,
  "Probe::Perl" => "0.01",
  "Scalar::Util" => "1.02",
  "Test::Builder" => 0,
  "Test::More" => 0,
  "Text::Balanced" => 0,
  "lib" => 0,
  "strict" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
